File: Week4/MVVMArchitectureComponents/app/src/main/java/com/yeah/ruisu/mvvmarchitecturecomponents/AppController.java
package com.yeah.ruisu.mvvmarchitecturecomponents;

import android.app.Application;

import com.yeah.ruisu.mvvmarchitecturecomponents.di.DaggerMainComponent;
import com.yeah.ruisu.mvvmarchitecturecomponents.di.MainComponent;
import com.yeah.ruisu.mvvmarchitecturecomponents.di.MainModule;

public class AppController extends Application
{
    private MainComponent MyMainComponent;

    public MainComponent getAppComponent()
    {
        if (MyMainComponent == null)
        {
            MyMainComponent = DaggerMainComponent.builder()
                                                .mainModule(new MainModule(this))
                                                .build();
        }
        return MyMainComponent;
    }
}

File: Reference/DeltaChallenge-master/DeltaChallenge-master/app/src/main/java/com/example/admin/deltachallenge/AppController.kt
package com.example.admin.deltachallenge

import android.app.Application
import com.example.admin.deltachallenge.di.application.ApplicationComponent
import com.example.admin.deltachallenge.di.application.ApplicationModule
import com.example.admin.deltachallenge.di.application.DaggerApplicationComponent

class AppController : Application() {

    private lateinit var applicationComponent: ApplicationComponent

    fun getComponent(): ApplicationComponent {
        return if (this::applicationComponent.isInitialized) applicationComponent else createApplicationComponent()
    }

    private fun createApplicationComponent(): ApplicationComponent {
        applicationComponent = DaggerApplicationComponent.builder().applicationModule(ApplicationModule(this@AppController)).build()
        return applicationComponent
    }
}